lib/Transforms/ConvertSecretForToStaticFor/ConvertSecretForToStaticFor.cpp

```diff
@@ -0,0 +1,208 @@
+
+namespace mlir {
+namespace heir {
+
+#define GEN_PASS_DEF_CONVERTSECRETFORTOSTATICFOR
+#include "lib/Transforms/ConvertSecretForToStaticFor/ConvertSecretForToStaticFor.h.inc"
+
+struct SecretForToStaticForConversion : OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern<scf::ForOp>::OpRewritePattern;
+
+ public:
+  SecretForToStaticForConversion(DataFlowSolver *solver, MLIRContext *context)
+      : OpRewritePattern(context), solver(solver){};
+
+  LogicalResult matchAndRewrite(scf::ForOp forOp,
+                                PatternRewriter &rewriter) const override {
+    auto *lowerBoundSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(forOp.getLowerBound());
+
+    auto *upperBoundSecretnessLattice =
+        solver->lookupState<SecretnessLattice>(forOp.getUpperBound());
+
+    // Note: Secretness lattice being null implies that the bounds were created
+    // by the first run of this pass.
+    // If the code has already been transformed, return
+    if (!lowerBoundSecretnessLattice && !upperBoundSecretnessLattice)
+      return failure();
+
+    // Get secretness state of the lower and upper bounds
+    bool isLowerBoundSecret =
+        lowerBoundSecretnessLattice &&
+        lowerBoundSecretnessLattice->getValue().getSecretness();
+    bool isUpperBoundSecret =
+        upperBoundSecretnessLattice &&
+        upperBoundSecretnessLattice->getValue().getSecretness();
+
+    // If both bounds are non-secret constants, return
+    if (!isLowerBoundSecret && !isUpperBoundSecret) return failure();
+
+    int newLowerBound, newUpperBound;
+
+    if (isLowerBoundSecret) {
+      // If static lower bound is not provided, emit a warning and return
+      // failure
+      if (!forOp->getAttrOfType<IntegerAttr>("lower")) {
+        InFlightDiagnostic diag = mlir::emitWarning(
+            forOp.getLoc(),
+            "Cannot convert secret scf.for to static affine.for "
+            "since a static lower bound attribute has not been provided:");
+
+        return failure();
+      }
+      // If lowerBound is secret, get value from "lower" attribute
+      newLowerBound = forOp->getAttrOfType<IntegerAttr>("lower").getInt();
+    }
+
+    if (isUpperBoundSecret) {
+      // If static upper bound is not provided, emit a warning and return
+      // failure
+      if (!forOp->getAttrOfType<IntegerAttr>("upper")) {
+        InFlightDiagnostic diag = mlir::emitWarning(
+            forOp.getLoc(),
+            "Cannot convert secret scf.for to static affine.for "
+            "since a static upper bound attribute has not been provided:");
+
+        return failure();
+      }
+      // If upperBound is secret, get value from "upper" attribute
+      newUpperBound = forOp->getAttrOfType<IntegerAttr>("upper").getInt();
+    }
+
+    ImplicitLocOpBuilder builder(forOp->getLoc(), rewriter);
+
+    auto newForOp = builder.create<affine::AffineForOp>(
```

Comment:
I assume that changes from scf.for to affine.for for the replcaement loop now "fixes" that issue because the new affine for loops' bound operands are guaranteed to not be in the secretness lattice. 

So I guess something happens when replacing with the original op that the original operands are still sometimes in that lattice?

Could you add a quick comment here about why you used AffineForOp (or perhaps link the new issue about it)?

Reply:
Replacing the secret-dependent for-loop with an affine.for loop was the original plan for this transformation for the program to benefit from other affine transformations. 

Although removing the original operands from the lattice "fixes" the redundant application of the transformation, I am not sure why but without your fix, requesting the secretness of the newly created forOp's upperBound returns a 1 (i.e., it's a secret), which is quite strange given that the new forOp's upperBound was just created. Your proposed patch works great but I am not sure why the new upperBound has any relation with the upperBound from the originalOp?

Reply:
Oh, got it! Makes sense to use affine for all the upstream fusion / normalization passes.

> I am not sure why the new upperBound has any relation with the upperBound of the originalOp?

Totally agree! It might be worth seeing if that's a bug in upstream MLIR's analyses by writing a really simple reproducer but I suspect we probably can't right now and in the meantime we'll keep an eye out if something like this reappears.